extensions/pyRevitDevTools.extension/pyRevitDev.tab/Debug.panel/Unit Tests.pulldown/Test Family API.pushbutton/script.py

```diff
@@ -0,0 +1,156 @@
+TYPES_TO_ADD = {"A": 1.0, "B": 2.0, "C": 3.0, "D": 4.0}
+LOG = script.get_logger()
+
+
+def _purge_family():
+    """Remove the temporary family from the document."""
+    to_delete = None
+    collector = DB.FilteredElementCollector(revit.doc).OfClass(DB.Family)
+    for family_element in collector.GetElementIterator():
+        if family_element.Name == TEMP_FAMILY_NAME:
+            to_delete = family_element.Id
+            break
+    if to_delete is not None:
+        message = "Purge {0} family".format(TEMP_FAMILY_NAME)
+        with Transaction(message, doc=revit.doc):
+            revit.doc.Delete(to_delete)
+            LOG.info(message)
+
+
+def _family_type_count(family_id):
+    """
+    Return the number of types in a loaded family.
+    """
+    if isinstance(family_id, int):
+        family_id = DB.ElementId(family_id)
+    elif not isinstance(family_id, DB.ElementId):
+        raise TypeError()
+    family = revit.doc.GetElement(family_id)
+    return len(list(family.GetFamilySymbolIds()))
+
+
+def _add_types(family_doc):
+    """
+    Create new types in the test family, and assign values to the type
+    parameter.
+    """
+    if not isinstance(family_doc, DB.Document):
+        raise TypeError()
+    if not family_doc.IsFamilyDocument:
+        raise ValueError()
+    manager = family_doc.FamilyManager
+    parameter = manager.get_Parameter(TEST_PARAMETER_NAME)
+    with Transaction("Add Family Types", doc=family_doc):
+        for k, v in TYPES_TO_ADD.items():
+            manager.NewType(k)
+            manager.Set(parameter, v)
+
+
+def _check_type_parameter_values(family_id):
+    """
+    Check the family types and parameter values in the loaded test family.
+    """
+    if isinstance(family_id, int):
+        family_id = DB.ElementId(family_id)
+    elif not isinstance(family_id, DB.ElementId):
+        raise TypeError()
+    family_element = revit.doc.GetElement(family_id)
+    for type_id in family_element.GetFamilySymbolIds():
+        type_element = revit.doc.GetElement(type_id)
+        type_name = type_element.LookupParameter("Type Name").AsString()
+        if type_name == TEMP_FAMILY_NAME:
+            continue
+        if type_name not in TYPES_TO_ADD:
+            raise KeyError("Unexpected Type Name: " + type_name)
+        parameter = type_element.LookupParameter(TEST_PARAMETER_NAME)
+        if parameter is None:
+            raise TypeError("Missing Type Parameter: " + TEST_PARAMETER_NAME)
+        if abs(parameter.AsDouble() - TYPES_TO_ADD[type_name]) > 0.0000001:
+            raise ValueError("Unexpected value in type parameter.")
+
+
+def test_family_api():
+    """Exercise the Revit family creation API."""
+    _purge_family()
+    family_template_path = HOST_APP.app.FamilyTemplatePath
+    assert os.path.isdir(family_template_path)
+    family_template = os.path.normpath(
+        os.path.join(
+            family_template_path,
+            r"English-Imperial\Annotations\Generic Annotation.rft",
```

Comment:
maybe adding a forms.alert() stating that it is required to run on Revit ENU and 2022 and over would help, OR, taking care of every case